New module sale_order_authorized_users

LeartS opened the pull request:
Let Admin (or any user in the `base.group_erp_manager` group) choose, for each sale order,
which users will be able to access and see it. Any other user won't be able to see it.
If no users are set, the sale order has normal permissions.
The field to set allowed users will be visibile only to admin, which makes it
possible to make the users unaware of this feature.

Things that could be improved:
- Tests?
- create security rule also for `sale_order_line` if it's not automatic.

It's not 100% secure at the moment, but probably good enough for "normal" users.

LeartS committed:
New module sale_order_authorized_users

LeartS committed:
[FIX] I bow before PEP8.

LeartS commented:
Build fails on main 7.0 branch and not because of my PR, branch status is green (when it should be red) because of https://github.com/OCA/maintainer-quality-tools/issues/60

pedrobaeza commented:
Can you rebase to see if the problem still remains?

LeartS commented:
@pedrobaeza the 7.0 branch is currently red (as the mentioned issue has been fixed), if I rebase now won't it fail just because the main branch is failing?

pedrobaeza commented:
Yeah, I see. I thought it was fixed already. Let's wait then that the upstream problem is fixed to do the rebase.

Thanks.

guewen commented on sale_order_authorized_users/sale.py:
```diff
@@ -0,0 +1,35 @@
+            'res.users',
+            'sale_order_res_users_rel',
+            'sale_order_id',
+            'user_id',
```
Expliciting the 2 relation fields and name of the relation table is useless, they will be generated by the ORM.

guewen commented on sale_order_authorized_users/sale.py:
```diff
@@ -0,0 +1,35 @@
+            'sale_order_res_users_rel',
+            'sale_order_id',
+            'user_id',
+            'Allowed Users',
```
Using the named parameter is cleaner, so it ends up to:

```
    'allowed_users_ids': fields.many2many(
        'res.users',
        string='Allowed Users',
        groups='base.group_erp_manager'
    ),
```

guewen commented on sale_order_authorized_users/security/hide_sale_orders_security.xml:
```diff
@@ -0,0 +1,13 @@
+		</record>
+
+	</data>
+</openerp>
```
Could use spaces (preferably 2 for xml files) instead of tabs?

guewen commented:
Apart the few line comments, seems good to me

LeartS committed:
[IMP] Field definition and whitespace fixes, under @guewen suggestions.

 * Convert tabs to spaces
 * 2 columns indentation for xml files
 * Named parameters and avoid unnecessary params in many2many field definition.

guewen commented:
Thanks

The build is red, but this is because of https://github.com/OCA/OCB/issues/37

So :+1:

lepistone commented on sale_order_authorized_users/sale.py:
```diff
@@ -0,0 +1,32 @@
+        'allowed_users_ids': fields.many2many(
+            'res.users',
+            string='Allowed Users',
+            groups='base.group_erp_manager',
```
can you add a help string saying that if left empty, normal rules apply?

lepistone commented on sale_order_authorized_users/security/hide_sale_orders_security.xml:
```diff
@@ -0,0 +1,13 @@
+      <field name="global" eval="True" />
+      <field name="domain_force">['|',('allowed_users_ids','in',user.id),('allowed_users_ids','=',False)]</field>
+    </record>
+
```
wouldn't it be straightforward to add a similar rule for lines just in case? like order_id.allowed_users_ids in user.id

lepistone commented on sale_order_authorized_users/__openerp__.py:
```diff
@@ -0,0 +1,47 @@
+======================================
+
+Let Admin (or any user in the base.group_erp_manager) choose,
+for each sale order,w hich users will be able to access and see it.
```
typo

lepistone commented:
Not blocking for that, but tests on user rights can be done, even though there are not many around. I'm happy to approve that once my minor remarks are addressed

LeartS committed:
Fix typo, add field help string, change manifest author.

LeartS committed:
Add security rule for sale order lines.

LeartS committed:
Add simple YAML test.

LeartS commented:
Thanks @lepistone, I updated the PR according to your suggestions and also created a simple YAML test. (my first, don't eat me if there are some errors!)

lepistone commented:
excellent @LeartS!

in fact, now you could add another case that tries to create an order with no authorized users.

LeartS commented:
@eLBati, why the needs fixing tag? Because I've not implemented the last test case suggested by @lepistone ?

eLBati commented:
Hello @LeartS. Yes, because of the missing test case

LeartS committed:
Add test case with sale order without authorized_users.

pedrobaeza commented:
One functional question: don't you think it's better to let any sale manager to decide authorized users instead the admin (usually a tech profile)?

LeartS commented:
@pedrobaeza I made this module because one of my customers asked me for a way to create "secret orders", only visible to people he specifically chooses, something separate and that takes precedence over the normal sale permissions schema. That's why I used `base.erp_manager` (admin).

If I used sale managers, all of them would see the `authorized_users` field and the module would be useless to hide a sale order from everyone or to some of them, for example. The idea is that nobody should even know there is an additional system to hide sale orders in place.

What I could do though, is create a new group (like `sale.secret_orders`) so that the admin can put in there users he want to be able to see and edit the `the authorized_users` field.

nhomar commented:
@LeartS I am agreed you should create a new group as you propose, that's the correct way to achieve that, base.erp_manager is a group for technical purpose.

LeartS committed:
Add group group_sale_secret_orders and use that instead of base.erp_manager.

gurneyalex commented:
The code looks fine. 

I'm not sure there is enough need for the functionality to warrant inclusion in OCA/sale-workflow. This sounds very customer specific to me. 

Edit: :+0:

LeartS commented:
Thanks for the review. If you guys think it's not a requested enough feature, I will put it up on my GitHub account, no problem.

pedrobaeza commented:
With the change I proposed, I think it's a nice to have feature, because it's very usual than inside a company you need to restrict sale orders/quotations to certain users. Granularity offered by OpenERP is not enough in these cases: your orders, sales team orders, or all.

LeartS commented:
@pedrobaeza I alread added the specific group, did you see?

Should I make that users in the secret orders group are automatically users of sale manager group?

lepistone commented:
that sounds like a good idea @LeartS . I approve the code, no opinion on it being too specific or not. Let's see if @gurneyalex is convinced by @pedrobaeza :smile_cat:

LeartS committed:
Sale / secret orders inherits from Sales manager.

LeartS commented:
Done. I'm not sure if it's ok to imply only sales manager or I should also include all the other sales groups.

pedrobaeza commented:
Thanks, for me it's :+1: .

@gurneyalex, features offered by this module is enough to justify it, so please reconsider you position.

Regards.

dcorio commented:
:+1: 

i would really like to see this applied (this way or another) to all the models, by default as a core feature.

LeartS committed:
Change description to reflect implementation changes, small xml fixes.

LeartS committed:
Fix description: you *must* set yourself as allowed or it won't work (unless you're working as admin).

LeartS commented:
I've changed the description in the manifest file to reflect the usage of the new group instead of `base.erp_manager`. I've also clarified (I hope) that the usual rules still apply, setting an user as allowed **doesn't** necessarily mean he will be able to see the order.

One "inconvenience" of using the custom made group is that if an user in the Secret orders group adds allowed users to a sale order, he must also explicitly add himself or he won't be able to make the change. I've also added this gotcha in the description.

LeartS committed:
Fix PEP8.

LeartS commented:
So, any chance of getting out of the limbo? :cloud:

lepistone commented:
hurm, @gurneyalex is against. I think you can close this if you're still of that idea Alexandre.

gurneyalex commented:
ok... I'll abstain... please disregard my previous vote and merge.

jgrandguillaume commented:
Hi,

Thanks for the contrib. I tempted to be inline with @gurneyalex here, may be it's a bit too specific. But as you manage to take into account all remarks and your module has a test, I think you deserve to have this module here. So I'm :+1: 

Regards,

Joël
